File: packages/sdk-python/kangopenbanking/qr.py
"""KOB QR Merchant Directory helpers (v4.31.x).

Public, unauthenticated. Cursor-paginated auto-fetch with a 5-minute
in-process cache so a partner virtual-card app can call ``qr.directory.list()``
on every scan without thrash.
"""
from __future__ import annotations

import json
import time
import urllib.parse
import urllib.request
from typing import Any, Dict, List, Optional

FN_BASE = "https://wdzkzeahdtxlynetndqw.supabase.co/functions/v1"
TTL_S = 5 * 60

_cache: Dict[str, Any] = {"key": None, "ts": 0.0, "data": None}


def _http_get_json(url: str) -> Dict[str, Any]:
    req = urllib.request.Request(url, headers={"Accept": "application/json"})
    with urllib.request.urlopen(req, timeout=15) as r:
        if r.status >= 400:
            raise RuntimeError(f"KOB QR error: HTTP {r.status}")
        return json.loads(r.read().decode("utf-8"))


def _fetch_all(country: Optional[str], category: Optional[str], hard_cap: int) -> List[Dict[str, Any]]:
    out: List[Dict[str, Any]] = []
    cursor: Optional[str] = None
    while len(out) < hard_cap:
        params = {"limit": "100"}
        if country:
            params["country"] = country
        if category:
            params["category"] = category
        if cursor:
            params["cursor"] = cursor
        url = f"{FN_BASE}/merchants-qr-directory?{urllib.parse.urlencode(params)}"
        page = _http_get_json(url)
        out.extend(page.get("data") or [])
        if not page.get("has_more") or not page.get("next_cursor"):
            break
        cursor = page["next_cursor"]
    return out


class _Directory:
    def list(self, country: Optional[str] = None, category: Optional[str] = None,
             hard_cap: int = 1000) -> List[Dict[str, Any]]:
        """Return the full active merchant directory. Cached for 5 minutes."""
        key = f"{country}|{category}|{hard_cap}"
        if _cache["key"] == key and (time.time() - _cache["ts"]) < TTL_S and _cache["data"] is not None:
            return _cache["data"]
        data = _fetch_all(country, category, hard_cap)
        _cache.update(key=key, ts=time.time(), data=data)
        return data

    def sync(self, **kwargs) -> List[Dict[str, Any]]:
        """Force-refresh ignoring the cache."""
        _cache["key"] = None
        return self.list(**kwargs)

    def by_id(self, **kwargs) -> Dict[str, Dict[str, Any]]:
        return {m["merchant_id"]: m for m in self.list(**kwargs)}


class _Merchant:
    def get(self, merchant_id: str, amount: Optional[str] = None, ref: Optional[str] = None) -> Dict[str, Any]:
        params = {"id": merchant_id}
        if amount:
            params["amount"] = amount
        if ref:
            params["ref"] = ref
        url = f"{FN_BASE}/merchants-qr-get?{urllib.parse.urlencode(params)}"
        return _http_get_json(url)


class _QR:
    directory = _Directory()
    merchant = _Merchant()


qr = _QR()

File: packages/sdk-node/src/qr.ts
// QR Merchant Directory helper (v4.31.x) — public, unauthenticated.
// Cursor-paginated auto-fetch with a 5-minute in-memory cache so a partner
// virtual-card app can call `directory.list()` on every scan without thrash.

const FN_BASE = 'https://wdzkzeahdtxlynetndqw.supabase.co/functions/v1';
const TTL_MS = 5 * 60 * 1000;

export interface QRDirectoryMerchant {
  merchant_id: string;
  name: string;
  environment: string;
  status: string;
  mcc: string | null;
  country: string | null;
  logo_url: string | null;
  verified: boolean;
  created_at: string;
}

export interface QRDirectoryFilters {
  country?: string;
  category?: string;
  hardCap?: number; // default 1000
}

let cache: { data: QRDirectoryMerchant[]; ts: number; key: string } | null = null;

async function fetchAll(filters: QRDirectoryFilters): Promise<QRDirectoryMerchant[]> {
  const cap = filters.hardCap ?? 1000;
  const out: QRDirectoryMerchant[] = [];
  let cursor: string | null = null;
  while (out.length < cap) {
    const url = new URL(`${FN_BASE}/merchants-qr-directory`);
    url.searchParams.set('limit', '100');
    if (filters.country) url.searchParams.set('country', filters.country);
    if (filters.category) url.searchParams.set('category', filters.category);
    if (cursor) url.searchParams.set('cursor', cursor);
    const res = await fetch(url.toString(), { headers: { Accept: 'application/json' } });
    if (!res.ok) throw new Error(`KOB QR directory error: HTTP ${res.status}`);
    const json = await res.json() as { data: QRDirectoryMerchant[]; has_more?: boolean; next_cursor?: string | null };
    out.push(...(json.data ?? []));
    if (!json.has_more || !json.next_cursor) break;
    cursor = json.next_cursor;
  }
  return out;
}

export const qr = {
  directory: {
    /** Returns the full active merchant directory. Cached for 5 minutes. */
    async list(filters: QRDirectoryFilters = {}): Promise<QRDirectoryMerchant[]> {
      const key = JSON.stringify(filters);
      if (cache && cache.key === key && Date.now() - cache.ts < TTL_MS) return cache.data;
      const data = await fetchAll(filters);
      cache = { data, ts: Date.now(), key };
      return data;
    },
    /** Force-refresh ignoring the cache. */
    async sync(filters: QRDirectoryFilters = {}): Promise<QRDirectoryMerchant[]> {
      cache = null;
      return this.list(filters);
    },
    /** Convenience map for O(1) merchant_id lookups after a scan. */
    async byId(filters: QRDirectoryFilters = {}): Promise<Map<string, QRDirectoryMerchant>> {
      const list = await this.list(filters);
      return new Map(list.map((m) => [m.merchant_id, m]));
    },
  },
  merchant: {
    async get(merchantId: string, opts: { amount?: string; ref?: string } = {}) {
      const url = new URL(`${FN_BASE}/merchants-qr-get`);
      url.searchParams.set('id', merchantId);
      if (opts.amount) url.searchParams.set('amount', opts.amount);
      if (opts.ref) url.searchParams.set('ref', opts.ref);
      const res = await fetch(url.toString(), { headers: { Accept: 'application/json' } });
      if (!res.ok) throw new Error(`KOB merchant QR error: HTTP ${res.status}`);
      return res.json();
    },
  },
};
